### PR TITLE
Embargoes and Leases from CurationConcerns

### LINKED ISSUE #792: Pull in EmbargosController from CurationConcerns
Currently there's no way to move embargoed items out of embargo.

### PR DESCRIPTION
**Pull request by darrenleeweber**
Fixes #792 and related to https://github.com/samvera-labs/hyku/issues/1184

Changes proposed in this pull request:
* copy everything for embargoes and leases from CurationConcerns
  * adapt everything to the `Hyrax` namespace
  * some of the required content is already in hyrax

TODO:
- [x] hook up routes
- [x] find and copy over any additional helpers, views, etc.
- [x] find and copy over all the specs and get them working
- [x] get the feature specs working
- [x] cleanup commits

@samvera-labs/hyrax-code-reviewers

### CODE REVIEW

**Comment by darrenleeweber** [comment]
2 spec failures:
```
Failures:

  1) embargo creating an embargoed object can be created, displayed and updated

     Failure/Error: click_link 'New Generic Work'

     Capybara::ElementNotFound:
       Unable to find link "New Generic Work"
     # ./spec/features/embargo_spec.rb:15:in `block (3 levels) in <top (required)>'

  2) Hyrax::EmbargoesController#destroy when I have permission to edit the object that has files deactivates embargo and checks to see if we want to copy the visibility to files

     Failure/Error: redirect_to confirm_curation_concerns_permission_path(curation_concern)

     NoMethodError:
       undefined method `confirm_curation_concerns_permission_path' for #<Hyrax::EmbargoesController:0x00000012f2d430>
     # ./app/controllers/concerns/hyrax/embargoes_controller_behavior.rb:16:in `destroy'
     # ./spec/controllers/hyrax/embargoes_controller_spec.rb:71:in `block (5 levels) in <top (required)>'
```

**Comment by darrenleeweber** [comment]
Embargo and Lease seem to go together, so added Leases.  But, some specs are still failing.

**Comment by darrenleeweber** [comment]
Marked some feature specs as pending, i.e.
- https://github.com/samvera-labs/hyrax/blob/embargos-controller-etc/spec/features/embargo_spec.rb#L14-L36
- https://github.com/samvera-labs/hyrax/blob/embargos-controller-etc/spec/features/lease_spec.rb#L10-L34

File: app/helpers/batch_edits_helper.rb [hunk 1]
```diff
@@ -1,14 +0,0 @@
-# View Helpers for Hydra Batch Edit functionality
-module BatchEditsHelper
```
Comment (jcoyne) [inline app/helpers/batch_edits_helper.rb]: Why did you move this helper to the Hyrax namespace?  Could this move have gone into a separate PR?

File: app/helpers/hyrax/batch_select_helper.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# View Helpers for Hydra Batch Edit functionality
+module Hyrax
+  module BatchSelectHelper
+    # Displays the button to select/deselect items for your batch.  Call this in the index
+    # partial that's rendered for each search result.
+    # @param [Hash] document the Hash (aka Solr hit) for one Solr document
+    def button_for_add_to_batch(document)
```
Comment (jcoyne) [inline app/helpers/hyrax/batch_select_helper.rb]: I don't ever see this method used?  Do we need this module?

File: app/views/hyrax/embargoes/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<h1>Manage Embargoes</h1>
```
Comment (jcoyne) [inline app/views/hyrax/embargoes/index.html.erb]: Should these be adapted to look like our current styles, or is that something you want to take on in a separate PR?

File: app/views/hyrax/leases/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<h1>Manage Leases</h1>
```
Comment (jcoyne) [inline app/views/hyrax/leases/index.html.erb]: Should these be adapted to look like our current styles, or is that something you want to take on in a separate PR?

**Comment by darrenleeweber** [comment]
Note that the last commit (https://github.com/samvera-labs/hyrax/pull/1042/commits/d66ffc150d10c1c253915195025c66c2845cc21b) moves some files around to consolidate things in the hyrax folders; not sure if that's OK.

File: lib/hyrax/rails/routes.rb [hunk 5]
```diff
@@ -46,20 +46,6 @@ def curation_concerns_basic_routes(&block)
       end
     end
 
-    # kmr added :show to make tests pass
-    def curation_concerns_embargo_management
```
Comment (jcoyne) [inline lib/hyrax/rails/routes.rb]: Do we need to deprecate this method?

File: app/helpers/hyrax/batch_select_helper.rb (hunk 2 above)
Comment (darrenleeweber) [inline app/helpers/hyrax/batch_select_helper.rb]: Discovered that using `render` in some views works and I _guess_ the render call is now _sorta_ using the controller module, which now includes this module.  For some reason, the views were not _finding_ this module/method automatically and I noticed that the batch edit modules have some explicit helper includes, so I followed that pattern for the batch select.

File: app/views/hyrax/embargoes/index.html.erb (hunk 3 above)
Comment (darrenleeweber) [inline app/views/hyrax/embargoes/index.html.erb]: If it's not much work, let's do it here in this PR.  Is there a good/related view style to copy/follow?

File: app/helpers/batch_edits_helper.rb (hunk 1 above)
Comment (darrenleeweber) [inline app/helpers/batch_edits_helper.rb]: Good question.  Perhaps the answer depends on an answer to the opposite question, i.e., why was it NOT in the hyrax namespace?  It could be a separate PR, but this PR is adding what seems to be related batch-select code, which is working when it's in the hyrax namespace.  To be consistent, it looks like the batch-* stuff can live in the hyrax namespace.  But, I dunno, /shrug

File: app/views/hyrax/leases/index.html.erb (hunk 4 above)
Comment (darrenleeweber) [inline app/views/hyrax/leases/index.html.erb]: Whatever is done for embargo can be done for leases - there _might_ be nearly identical partials.

File: lib/hyrax/rails/routes.rb (hunk 5 above)
Comment (darrenleeweber) [inline lib/hyrax/rails/routes.rb]: /shrug, but, if so, how is that to be done?

Comment (jcoyne) [inline lib/hyrax/rails/routes.rb]: keep the method, make it do nothing and have:

```ruby
Deprecation.warn(self, "curation_concerns_embargo_management is deprecated and has no effect. It will be removed in Hyrax 3.0")
```

File: app/views/hyrax/embargoes/index.html.erb (hunk 3 above)
Comment (jcoyne) [inline app/views/hyrax/embargoes/index.html.erb]: Something like this: https://github.com/samvera-labs/hyrax/blob/master/app/views/hyrax/admin/admin_sets/index.html.erb

File: app/helpers/hyrax/batch_select_helper.rb (hunk 2 above)
Comment (jcoyne) [inline app/helpers/hyrax/batch_select_helper.rb]: You stopped calling this method when you made this change: https://github.com/samvera-labs/hyrax/pull/1042/commits/d66ffc150d10c1c253915195025c66c2845cc21b#diff-4e8fdc051d085b4be70f397bbccffb11L26 So please remove this module.

File: app/helpers/batch_edits_helper.rb (hunk 1 above)
Comment (jcoyne) [inline app/helpers/batch_edits_helper.rb]: It's much easier to approve smaller PRs, so if you can make this change against master without all the other changes in this PR that would be preferable.

File: app/helpers/hyrax/batch_select_helper.rb (hunk 2 above)
Comment (darrenleeweber) [inline app/helpers/hyrax/batch_select_helper.rb]: Pushed up a commit to remove this module and related includes.

File: app/helpers/batch_edits_helper.rb (hunk 1 above)
Comment (darrenleeweber) [inline app/helpers/batch_edits_helper.rb]: Trying to move this to https://github.com/samvera-labs/hyrax/pull/1114

File: lib/hyrax/rails/routes.rb (hunk 5 above)
Comment (darrenleeweber) [inline lib/hyrax/rails/routes.rb]: Done

File: app/views/hyrax/embargoes/index.html.erb (hunk 3 above)
Comment (darrenleeweber) [inline app/views/hyrax/embargoes/index.html.erb]: Attempted, using best guess, in 0d7b622

File: app/views/hyrax/leases/index.html.erb (hunk 4 above)
Comment (darrenleeweber) [inline app/views/hyrax/leases/index.html.erb]: Attempted, using best guess, in 0d7b622

**Comment by darrenleeweber** [comment]
Locally, I see specs passing on this
```
Finished in 16 minutes 17 seconds (files took 13.44 seconds to load)
2296 examples, 0 failures, 12 pending
```

**Comment by darrenleeweber** [comment]
> I'm not sure why batch_select/_add_button.html.erb was moved in this PR. Could that have been a separate PR?

Also moved to an additional commit in #1114

File: lib/hyrax/rails/routes.rb [hunk 6]
```diff
@@ -46,18 +46,9 @@ def curation_concerns_basic_routes(&block)
       end
     end
 
-    # kmr added :show to make tests pass
+    Deprecation.warn(self, "#curation_concerns_embargo_management is deprecated and has no effect. It will be removed in Hyrax 3.0")
```
Comment (jcoyne) [inline lib/hyrax/rails/routes.rb]: This has to be inside the body of the `curation_concerns_embargo_management` method.

File: spec/support/shared/shared_examples_is_embargoable.rb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+shared_examples 'is_embargoable' do
```
Comment (jcoyne) [inline spec/support/shared/shared_examples_is_embargoable.rb]: This shared example is never used. Can we remove this file?

File: app/controllers/concerns/hyrax/embargoes_controller_behavior.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
+module Hyrax
+  module EmbargoesControllerBehavior
+    extend ActiveSupport::Concern
+    include Hyrax::ManagesEmbargoes
+    include Hyrax::Collections::AcceptsBatches
+
+    included do
+      # include the render_check_all view helper method
+      helper ::BatchEditsHelper
```
Comment (jcoyne) [inline app/controllers/concerns/hyrax/embargoes_controller_behavior.rb]: Can you say why this needed to be added? I'm not seeing `render_check_all` called anywhere in this PR.

File: app/controllers/concerns/hyrax/leases_controller_behavior.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
+module Hyrax
+  module LeasesControllerBehavior
+    extend ActiveSupport::Concern
+    include Hyrax::ManagesEmbargoes
+    include Hyrax::Collections::AcceptsBatches
+
+    included do
+      # include the render_check_all view helper method
+      helper ::BatchEditsHelper
```
Comment (jcoyne) [inline app/controllers/concerns/hyrax/leases_controller_behavior.rb]: Why is this necessary?

**Comment by darrenleeweber** [comment]
Rebased on master after merge of #1114 and cleaned up some commits.

File: spec/support/shared/shared_examples_is_embargoable.rb (hunk 7 above)
Comment (darrenleeweber) [inline spec/support/shared/shared_examples_is_embargoable.rb]: Yes, if specs pass without it and the specs marked as pending do not need it.

**Comment by darrenleeweber** [comment]
> If you want to, you could merge LeasesControllerBehavior into LeasesController and EmbargosControllerBehavior into EmbargosController. This change has already been made for other controllers in Hyrax. 

Pushed up this change.

**Comment by darrenleeweber** [comment]
Many spec failures locally, working to resolve them.

**Comment by darrenleeweber** [comment]
Specs are passing for me
```
Finished in 13 minutes 31 seconds (files took 10.79 seconds to load)
2299 examples, 0 failures, 13 pending
```

**Comment by jcoyne** [comment]
Looks good to me. This one is big, so can someone else also review and give the approval?

File: app/controllers/concerns/hyrax/manages_embargoes.rb [hunk 10]
```diff
@@ -0,0 +1,21 @@
+module Hyrax
+  module ManagesEmbargoes
+    extend ActiveSupport::Concern
+
+    included do
+      include Hyrax::ThemedLayoutController
+      with_themed_layout '1_column'
```
Comment (cbeer) [inline app/controllers/concerns/hyrax/manages_embargoes.rb]: I believe `with_themed_layout '1_column'` is the default these days, isn't it?

File: app/views/hyrax/embargoes/_list_active_embargoes.html.erb [hunk 11]
```diff
@@ -0,0 +1,16 @@
+<table class="embargoes table">
+  <tbody>
+  <tr>
+    <th>Type of Work</th><th>Title</th><th>Current Visibility</th><th>Embargo Release Date</th><th>Visibility will Change to</th>
```
Comment (cbeer) [inline app/views/hyrax/embargoes/_list_active_embargoes.html.erb]: Shouldn't all these text strings be i18n'ed?

File: spec/features/embargo_spec.rb [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+  describe 'creating an embargoed object' do
+    before do
+      visit '/'
+    end
+
+    let(:future_date) { 5.days.from_now }
+    let(:later_future_date) { 10.days.from_now }
+
+    # Pending revision to work in hyrax
+    xit 'can be created, displayed and updated', :workflow do
```
Comment (cbeer) [inline spec/features/embargo_spec.rb]: What revision is necessary? Should that be ticketed and at least left as a hint to the next person?

File: app/views/hyrax/embargoes/_list_active_embargoes.html.erb (hunk 11 above)
Comment (mjgiarlo) [inline app/views/hyrax/embargoes/_list_active_embargoes.html.erb]: :+1:, though would be happy for that to be handled in a follow-on PR.

File: spec/features/embargo_spec.rb (hunk 12 above)
Comment (mjgiarlo) [inline spec/features/embargo_spec.rb]: :+1: to creating an issue and linking it in via `# TODO: ...` comment here.

File: app/views/hyrax/embargoes/_list_active_embargoes.html.erb (hunk 11 above)
Comment (cbeer) [inline app/views/hyrax/embargoes/_list_active_embargoes.html.erb]: Then at the very least, I'd like to see that ticketed, although I think it'd be easier in the long run to at least extract these into an `en` i18n file, instead of having to dive through all our code later.

File: app/controllers/concerns/hyrax/manages_embargoes.rb (hunk 10 above)
Comment (mjgiarlo) [inline app/controllers/concerns/hyrax/manages_embargoes.rb]: Yes, I believe so; these two lines may not be necessary.

**Comment by mjgiarlo** [comment]
@hannahfrost @ggeisler After this PR is merged, we should link these pages into our current dashboard. Questions:

1. Where do we want embargoes and leases in the dashboard menu?
2. Who should see the embargo and leave menu items? (It's currently limited to repo admins, and they can see *all* embargoed and leased works. Maybe that's OK for an initial implementation, and we need another ticket for letting admin set managers see them, including making sure the lease and embargo controllers and views only show admin set managers works in their admin sets?)

File: app/views/hyrax/embargoes/_list_active_embargoes.html.erb (hunk 11 above)
Comment (mjgiarlo) [inline app/views/hyrax/embargoes/_list_active_embargoes.html.erb]: Agree, @cbeer -- I was thinking we could ask @darrenleeweber to take that on immediately following this PR so we wouldn't have to do said diving.

Comment (darrenleeweber) [inline app/views/hyrax/embargoes/_list_active_embargoes.html.erb]: -> https://github.com/samvera-labs/hyrax/issues/1126

File: spec/features/embargo_spec.rb (hunk 12 above)
Comment (darrenleeweber) [inline spec/features/embargo_spec.rb]: Noted in https://github.com/samvera-labs/hyrax/issues/1123 and a new squashed commit will contain the comment link in the code.

Comment (cbeer) [inline spec/features/embargo_spec.rb]: I was hoping #1123 would include information about why they were marked as pending and what work needs to be done first?

File: app/controllers/concerns/hyrax/manages_embargoes.rb (hunk 10 above)
Comment (darrenleeweber) [inline app/controllers/concerns/hyrax/manages_embargoes.rb]: Does this imply the install generator overrides this? i.e.
- https://github.com/samvera-labs/hyrax/blob/master/lib/generators/hyrax/install_generator.rb#L40

That generator _seems_ to apply this to a blacklight layout, so does that apply to this too?

Are you wanting both lines 6 & 7 removed or just line 7?

Comment (mjgiarlo) [inline app/controllers/concerns/hyrax/manages_embargoes.rb]: @darrenleeweber The install generator sets a default for Hyrax-based applications by adding that code to the app's `ApplicationController`. It'd be worth removing both lines and seeing what effect that has on the UI.

Comment (darrenleeweber) [inline app/controllers/concerns/hyrax/manages_embargoes.rb]: Done:  https://github.com/samvera-labs/hyrax/pull/1042/commits/4280e19649e0802bd9b2594c085daf458c5a6a76

**Comment by darrenleeweber** [comment]
It works for me:
```Finished in 13 minutes 51 seconds (files took 10.21 seconds to load)
2299 examples, 0 failures, 13 pending
```

I haven’t _seen_ this stuff in action in the UI yet.  Although specs are passing OK, the feature specs are a TODO item ( #1123 ) and might be better tackled when the dashboard integration is designed ( #1127 ).

**Comment by mjgiarlo** [comment]
Restarted the builds that failed because of two flappy feature specs.

**Comment by darrenleeweber** [comment]
The push, but not the pr, build is passing, despite restarting the pr-build several times over the last few workdays.  Maybe it has something to do with merging some of the other behavior->controller code?  The branch in PR 1042 has not been rebased on master since that work was merged.

File: app/views/hyrax/embargoes/_embargo_history.html.erb [hunk 13]
```diff
@@ -0,0 +1,7 @@
+<ul>
+  <% curation_concern.embargo_history.each do |entry| %>
+    <% unless entry.empty? %>
```
Comment (jeremyf) [inline app/views/hyrax/embargoes/_embargo_history.html.erb]: Prefer `if entry.present?` (I also don't believe that `nil` responds to `empty?`)

File: app/views/hyrax/embargoes/_list_deactivated_embargoes.html.erb [hunk 14]
```diff
@@ -0,0 +1,10 @@
+<% assets_with_deactivated_embargoes.each do |curation_concern| %>
+  <%= link_to curation_concern, edit_embargo_path(curation_concern) %> <%= visibility_badge(curation_concern.visibility) %>
+  <ul>
+    <% curation_concern.embargo_history.each do |entry| %>
+      <% unless entry.empty? %>
```
Comment (jeremyf) [inline app/views/hyrax/embargoes/_list_deactivated_embargoes.html.erb]: Prefer `if entry.present?` (unless statements can spiral into confusing things to parse; imagine adding another condition to this statement)

File: app/views/hyrax/embargoes/_list_expired_active_embargoes.html.erb [hunk 15]
```diff
@@ -0,0 +1,44 @@
+<% if assets_with_expired_embargoes.empty? %>
```
Comment (jeremyf) [inline app/views/hyrax/embargoes/_list_expired_active_embargoes.html.erb]: I think you may want to switch this to `if assets_with_expired_embaroes.present?` and switch the bodies of the `if` and `else` statements.

File: app/views/hyrax/embargoes/edit.html.erb [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+        <%= f.submit "Apply Embargo", class: 'btn btn-primary' %>
+      <% end %>
+      <%= link_to 'Cancel and manage all embargoes', embargoes_path, class: 'btn btn-default' %>
+      <%= link_to "Return to editing this #{curation_concern.human_readable_type}", [:edit, curation_concern], class: 'btn btn-default' %>
+    </div>
+  </div>
+<% end %>
+
+<h2>Past Embargoes</h2>
+<% if curation_concern.embargo_history.empty? %>
```
Comment (jeremyf) [inline app/views/hyrax/embargoes/edit.html.erb]: Perfer `if present?` and switch body of the `if` and `else` statement.

File: app/views/hyrax/leases/_lease_history.html.erb [hunk 17]
```diff
@@ -0,0 +1,7 @@
+<ul>
+  <% curation_concern.lease_history.each do |entry| %>
+    <% unless entry.empty? %>
```
Comment (jeremyf) [inline app/views/hyrax/leases/_lease_history.html.erb]: Prefer `if entry.present?`

File: app/views/hyrax/leases/_list_expired_active_leases.html.erb [hunk 18]
```diff
@@ -0,0 +1,43 @@
+<% if assets_with_expired_leases.empty? %>
```
Comment (jeremyf) [inline app/views/hyrax/leases/_list_expired_active_leases.html.erb]: Prefer `if present?` test (and switching `if` and `else`)

File: app/views/hyrax/leases/edit.html.erb [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+        <%= f.submit "Apply Lease", class: 'btn btn-primary' %>
+      <% end %>
+      <%= link_to 'Cancel and manage all leases', leases_path, class: 'btn btn-default' %>
+      <%= link_to "Return to editing this #{curation_concern.human_readable_type}", [:edit, curation_concern], class: 'btn btn-default' %>
+    </div>
+  </div>
+<% end %>
+
+<h2>Past Leases</h2>
+<% if curation_concern.lease_history.empty? %>
```
Comment (jeremyf) [inline app/views/hyrax/leases/edit.html.erb]: Prefer `if present?` test (and switching `if` and `else`)

**Comment by mjgiarlo** [comment]
Ah, good calls, @jeremyf. Friendly amendment: can you use `#any?` instead of `#present?` in any of those contexts? (You can call that method on `Enumerable` and `ActiveRecord::Relation`.)

**Comment by jcoyne** [comment]
@jeremyf I don't want those changes in this PR. This PR was to pull stuff from curation_concerns as faithfully as possible. These idiosyncrasies are present in CC and we should change them in a follow up PR.

File: spec/features/lease_spec.rb [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+  let(:user) { create(:user) }
+  before do
+    sign_in user
+  end
+  describe 'create a new leased object' do
+    let(:future_date) { 5.days.from_now }
+    let(:later_future_date) { 10.days.from_now }
+
+    # Pending revision to work in hyrax
+    # TODO: https://github.com/samvera-labs/hyrax/issues/1123
```
Comment (jcoyne) [inline spec/features/lease_spec.rb]: I don't think this is acceptable. We need to make the tests pass in order to merge this work in. I suspect the are not passing because the tests are written with the assumption that it's starting from an empty repository.  If you add `:clean_repo` to the spec metadata, that may fix it.

**Comment by mjgiarlo** [comment]
@darrenleeweber The feedback on the latest round of code review is good -- you've also been on this a while, so I'd like to do what I can to help expedite completion of this work (without sacrificing quality, of course). How would you feel about pairing with someone to finish this PR today?

File: app/views/hyrax/leases/_lease_history.html.erb (hunk 17 above)
Comment (atz) [inline app/views/hyrax/leases/_lease_history.html.erb]: Don't care on this (or similar nits).

**Comment by darrenleeweber** [comment]
@mjgiarlo - I'm not available to work on this at all this week  (conference Mon-Wed, vacay Thu-Fri).  If you like, someone else can take this over?  Otherwise, I can pick it up again next week.

**Comment by mjgiarlo** [comment]
@darrenleeweber Next week is A-OK.

**Comment by darrenleeweber** [comment]
Rebased on master, leaving specs to travis.

File: spec/features/lease_spec.rb (hunk 20 above)
Comment (darrenleeweber) [inline spec/features/lease_spec.rb]: OK pushed up this change in a WIP commit to see if it works out.  However, the feature specs may be entirely revised after the design work is done to integrate these features into hyrax.  If this spec-metadata addition does not work, I suggest the feature specs are re-written once the design work is done, unless there is a clear expectation that all these CC views _should_ be simply plug-n-play in hyrax (of course, this was not explicit in the requirements for this work).

**Comment by mjgiarlo** [comment]
@jeremyf @jcoyne's review reminds me that this work is primarily porting code from CurationConcerns, so I'm :-1: on the changes you suggest (though I generally favor the styles you're advocating). Given that, and how patiently @darrenleeweber has been waiting on this PR, I'm inclined to punt on the specific points from your review at this time. Thanks for the review -- always appreciated.

**Comment by darrenleeweber** [comment]
@jcoyne - the proposed change to add the `:clean_repo` to the feature specs did not resolve the failures, which are:
```
Failures:

  1) leases create a new leased object can be created, displayed and updated

     Failure/Error: click_link 'New Generic Work'
     
     Capybara::ElementNotFound:
       Unable to find link "New Generic Work"
     # ./spec/features/lease_spec.rb:15:in `block (3 levels) in <top (required)>'

  2) embargo creating an embargoed object can be created, displayed and updated

     Failure/Error: click_link 'New Generic Work'
     Capybara::ElementNotFound:

       Unable to find link "New Generic Work"
     # ./spec/features/embargo_spec.rb:18:in `block (3 levels) in <top (required)>'
```

@mjgiarlo - I thought we had general agreement that resolving these feature specs would be handled in a new PR that incorporates additional design work to include these features into hyrax and this work is ticketed already.  If that's the case, the last commit on this PR can be backed-out.

**Comment by mjgiarlo** [comment]
@darrenleeweber I'm fine with you taking up the feature specs in a PR that immediately follows this PR, yes, since the current UI doesn't yet expose the new views.

**Comment by mjgiarlo** [comment]
@jcoyne Can you be OK with :point_up: ?

**Comment by mjgiarlo** [comment]
@darrenleeweber after hearing both your and @jcoyne's concerns, I believe the best way forward is for the two of you to pair up early next week to march this across the finish line together. Thanks for your patience.

**Comment by darrenleeweber** [comment]
@mjgiarlo - would you like to mark the code-climate issue as a `wont fix` or something?

**Comment by darrenleeweber** [comment]
The travis-pr specs are failing on some file_set specs, might be resolved by https://github.com/samvera/hyrax/pull/1204

File: app/views/hyrax/embargoes/_list_deactivated_embargoes.html.erb (hunk 14 above)
Comment (atz) [inline app/views/hyrax/embargoes/_list_deactivated_embargoes.html.erb]: I find such difficulty negligible.  `unless` is particularly useful in guard clauses, e.g.: 
```ruby
next unless obj.condition? && obj.match(/regex/)
```
Where each conditional chunk then reflects an assertion that can be made about the object inside the loop.

Comment (atz) [inline app/views/hyrax/embargoes/_list_deactivated_embargoes.html.erb]: That being said, I agree that `nil` doesn't respond to `empty?` and there is no actual need to have this logic in the block.  Could just do: 
```erb
<% curation_concern.embargo_history.select(&:present?).each do |entry| %>
```
Not sure I would hold up a PR this big/old over it though.

File: app/views/hyrax/leases/_lease_history.html.erb (hunk 17 above)
Comment (atz) [inline app/views/hyrax/leases/_lease_history.html.erb]: OK, I care a little.

**Comment by darrenleeweber** [comment]
Pushed a commit squash - all the checks should be fine.
